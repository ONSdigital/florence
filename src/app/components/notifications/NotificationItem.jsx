import React from "react";
import PropTypes from "prop-types";
import clsx from "clsx";

const NotificationItem = props => (
    <li
        key={props.id + props.type}
        className={clsx("notifications__item", { visible: props.isVisible }, `notifications__item--${props.type}`)}
        data-testid={props.id + props.type}
    >
        {props.message}
        {props.buttons.map((button, i) => {
            return (
                <button key={`button-${i}`} className="notifications__button" onClick={button.onClick}>
                    {button.text}
                    {button.icon}
                </button>
            );
        })}
    </li>
);

NotificationItem.propTypes = {
    type: PropTypes.oneOf(["neutral", "warning", "positive"]).isRequired,
    id: PropTypes.string.isRequired,
    message: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
    isVisible: PropTypes.bool,
    buttons: PropTypes.arrayOf(
        PropTypes.shape({
            onClick: PropTypes.func.isRequired,
            text: PropTypes.string.isRequired,
            icon: PropTypes.element,
        })
    ),
};

export default NotificationItem;
