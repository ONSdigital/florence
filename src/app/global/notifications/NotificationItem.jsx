import React, { Component } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
    type: PropTypes.oneOf(["neutral", "warning"]).isRequired,
    message: PropTypes.string,
    id: PropTypes.string.isRequired,
    buttons: PropTypes.arrayOf(PropTypes.shape({
        onClick: PropTypes.func.isRequired,
        text: PropTypes.string.isRequired
    })),
    isVisible: PropTypes.bool
}

const defaultProps = {
    isVisible: false,
    buttons: []
}

class NotificationItem extends Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <li className={`notifications__item ${this.props.isVisible ? "visible" : ""} ${this.props.type === "warning" ? "notifications__item--warning" : ""}`}>
                {this.props.message}
                {this.props.buttons.map((button, index) => {
                    return ( 
                        <button 
                            key={index} 
                            className="notifications__button" 
                            onClick={button.onClick}
                        >
                            {button.text}
                        </button>
                    )
                })}
            </li>
        )
    }
}

NotificationItem.propTypes = propTypes;
NotificationItem.defaultProps = defaultProps;

export default NotificationItem;