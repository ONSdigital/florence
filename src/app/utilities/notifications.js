import { addNotification, removeNotification, toggleNotificationVisibility } from '../config/actions';
import { store } from '../config/store';

export default class notifications {

    /**
     * Updates state with new notification
     * @param {object} configuration of notification , as follows: 
     * {
     *      id [String]
     *      type [oneOf(["neutral", "warning"])]
     *      message [String]
     *      autoDismiss [Number]
     *      isDismissable [Bool]
     *      buttons [Array of objects]
     * }
     */
    static add(notification) {
        const config = {
            type: notification.type || "neutral",
            message: notification.message || "",
            id: notification.id || Date.now(),
            buttons: notification.buttons || [],
            isVisible: false
        }

        if (notification.autoDismiss && notification.autoDismiss > 0) {
            const timer = window.setTimeout(() => {
                window.clearTimeout(timer);
                this.remove(config.id);
            }, notification.autoDismiss);
        }

        if (notification.isDismissable) {
            config.buttons.push({
                text: "Close",
                onClick: function() {
                    notifications.remove(config.id)
                }
            });
        }

        store.dispatch(addNotification(config));
        
        // Set a timeout so browser doesn't try to render component without the animation
        const animationTimer = window.setTimeout(() => {
            store.dispatch(toggleNotificationVisibility(config.id));
            window.clearTimeout(animationTimer);
        }, 10);
    }

    /**
     * Removes notification from state
     * @param {string} notificationID 
     */
    static remove(notificationID) {
        store.dispatch(toggleNotificationVisibility(notificationID));

        // Set a timeout so browser doesn't try to render component without the animation
        const animationTimer = window.setTimeout(() => {
            store.dispatch(removeNotification(notificationID));
            window.clearTimeout(animationTimer);
        }, 30);
    }

}