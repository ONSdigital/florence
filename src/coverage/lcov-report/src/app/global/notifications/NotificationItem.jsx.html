<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/app/global/notifications/NotificationItem.jsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../../prettify.css" />
    <link rel="stylesheet" href="../../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../../index.html">All files</a> / <a href="index.html">src/app/global/notifications</a> NotificationItem.jsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>7/7</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">66.67% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>4/6</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>3/3</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>7/7</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { Component } from 'react';
import PropTypes from 'prop-types';
&nbsp;
const propTypes = {
    type: PropTypes.oneOf(["neutral", "warning", "positive"]).isRequired,
    message: PropTypes.string,
    id: PropTypes.string.isRequired,
    buttons: PropTypes.arrayOf(PropTypes.shape({
        onClick: PropTypes.func.isRequired,
        text: PropTypes.string.isRequired
    })),
    isVisible: PropTypes.bool
}
&nbsp;
const defaultProps = {
    isVisible: false,
    buttons: []
}
&nbsp;
class NotificationItem extends Component {
    constructor(props) {
        super(props);
    }
&nbsp;
    render() {
        return (
            &lt;li className={`notifications__item ${this.props.isVisible ? <span class="branch-0 cbranch-no" title="branch not covered" >"visible" </span>: ""} ${this.props.type === "warning" ? "notifications__item--warning" : ""} ${this.props.type === "positive" ? <span class="branch-0 cbranch-no" title="branch not covered" >"notifications__item--positive" </span>: ""}`}&gt;
                {this.props.message}
                {this.props.buttons.map((button, index) =&gt; {
                    return ( 
                        &lt;button 
                            key={index} 
                            className="notifications__button" 
                            onClick={button.onClick}
                        &gt;
                            {button.text}
                        &lt;/button&gt;
                    )
                })}
            &lt;/li&gt;
        )
    }
}
&nbsp;
NotificationItem.propTypes = propTypes;
NotificationItem.defaultProps = defaultProps;
&nbsp;
export default NotificationItem;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Aug 09 2017 09:26:50 GMT+0100 (BST)
</div>
</div>
<script src="../../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../../sorter.js"></script>
</body>
</html>
