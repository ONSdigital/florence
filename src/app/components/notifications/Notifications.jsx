import React from "react";
import PropTypes from "prop-types";
import NotificationItem from "./NotificationItem";

const Notifications = ({ notifications }) => {
    if (notifications.length === 0) return null;

    return (
        <ul className="notifications" data-testid="notifications">
            {notifications.map((notification, i) => (
                <NotificationItem key={notification + i} {...notification} />
            ))}
        </ul>
    );
};

Notifications.propTypes = {
    notifications: PropTypes.array,
};

export default Notifications;
