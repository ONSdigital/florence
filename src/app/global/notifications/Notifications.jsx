import React, { Component } from "react";
import PropTypes from "prop-types";

import NotificationItem from "./NotificationItem";

const propTypes = {
    notifications: PropTypes.arrayOf(PropTypes.object)
};

class Notifications extends Component {
    render() {
        return (
            <ul className="notifications">
                {this.props.notifications.map(notification => {
                    return <NotificationItem {...notification} key={notification.id} />;
                })}
            </ul>
        );
    }
}

Notifications.propTypes = propTypes;

export default Notifications;
