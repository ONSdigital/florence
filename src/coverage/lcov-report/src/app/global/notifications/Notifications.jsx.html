<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/app/global/notifications/Notifications.jsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../../prettify.css" />
    <link rel="stylesheet" href="../../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../../index.html">All files</a> / <a href="index.html">src/app/global/notifications</a> Notifications.jsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>4/4</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>4/4</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { Component } from 'react';
import PropTypes from 'prop-types';
&nbsp;
import NotificationItem from './NotificationItem';
&nbsp;
const propTypes = {
    notifications: PropTypes.arrayOf(PropTypes.object)
};
&nbsp;
class Notifications extends Component {
    render() {
        return (
            &lt;ul className="notifications"&gt;
                {this.props.notifications.map(notification =&gt; {
                    return &lt;NotificationItem {...notification} key={notification.id} /&gt;
                })}
            &lt;/ul&gt;
        )
    }
}
&nbsp;
Notifications.propTypes = propTypes;
&nbsp;
export default Notifications;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Aug 09 2017 09:26:50 GMT+0100 (BST)
</div>
</div>
<script src="../../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../../sorter.js"></script>
</body>
</html>
