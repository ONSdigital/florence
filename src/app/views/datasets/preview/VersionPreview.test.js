import React from 'react';
import {VersionPreview} from './VersionPreview';
import notifications from '../../../utilities/notifications';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

console.error = jest.fn();

jest.mock('../../../utilities/notifications', () => {
    return {
        add: jest.fn(() => {
            //
        })
    }
});
jest.mock('../../../utilities/url', () => {
    return {
        resolve: function() {
            //
        },
        parent: function() {}
    }
});
jest.mock('../../../utilities/api-clients/datasets', () => (
    {
        get: jest.fn(() => {
            return Promise.resolve({
                "id": "931a8a2a-0dc8-42b6-a884-7b6054ed3b68",
                "current": {
                    "collection_id": "fddffdfdaadf-e8ad17766a81bf589e76ef57d854945fdf0bfe546000228837aa70701506869c",
                    "id": "931a8a2a-0dc8-42b6-a884-7b6054ed3b68",
                    "links": {
                        "editions": {
                            "href": "http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions"
                        },
                        "latest_version": {
                            "id": "efcc4581-30b1-463b-b85b-2e2d85c4918b",
                            "href": "http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions/2016/versions/1"
                        },
                        "self": {
                            "href": "http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68"
                        }
                    },
                    "next_release": "pudding",
                    "publisher": {},
                    "qmi": {},
                    "state": "published",
                    "title": "CPI",
                    "uri": "/economy/inflationandpricesindices/datasets/consumerpriceindices"
                },
                "next": {
                    "collection_id": "fddffdfdaadf-e8ad17766a81bf589e76ef57d854945fdf0bfe546000228837aa70701506869c",
                    "id": "931a8a2a-0dc8-42b6-a884-7b6054ed3b68",
                    "links": {
                        "editions": {
                            "href": "http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions"
                        },
                        "latest_version": {
                            "id": "efcc4581-30b1-463b-b85b-2e2d85c4918b",
                            "href": "http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions/2016/versions/1"
                        },
                        "self": {
                            "href": "http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68"
                        }
                    },
                    "next_release": "pudding",
                    "publisher": {},
                    "qmi": {},
                    "state": "published",
                    "title": "CPI",
                    "uri": "/economy/inflationandpricesindices/datasets/consumerpriceindices"
                }
            })
        }).mockImplementationOnce(() => {
            return Promise.reject({status: 403});
        }),
        approveInstance: jest.fn().mockImplementation(() => {
            return Promise.resolve()
        })
    }
));

const mockEvent = {
    preventDefault: function() {}
}

const defaultProps = {
    dispatch: function(){},
    params: {
        datasetID: "abc123",
        edition: "2017",
        version: "2"
    }
}

test("Version preview matches stored snapshot", () => {
    const component = renderer.create(
        <VersionPreview {...defaultProps} />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test("On approval the state is updated to toggle disabling some actions during approval", async () => {
    const component = shallow(
        <VersionPreview {...defaultProps} />
    );

    expect(component.state("isApprovingVersion")).toBe(false);
    await component.instance().handleApproveSubmit(mockEvent);
    expect(component.state("isApprovingVersion")).toBe(true);
});

test("User shown a notification if an error occurs during approval", async () => {
    const component = shallow(
        <VersionPreview {...defaultProps} />
    );

    await component.instance().handleApproveSubmit(mockEvent);
    expect(notifications.add.mock.calls.length).toBe(1);
});

test("Dataset title updates after successful fetch from dataset API on mount", async () => {
    const component = shallow(
        <VersionPreview {...defaultProps} />
    );

    expect(component.state("datasetTitle")).toBe(null);
    await component.instance().componentWillMount();
    expect(component.update().state("datasetTitle")).toBe("CPI");
});