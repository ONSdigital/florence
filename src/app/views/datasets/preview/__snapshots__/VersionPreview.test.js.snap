// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Version preview matches stored snapshot 1`] = `
<div
  className="preview"
>
  <div
    className="preview__header grid grid--justify-center"
  >
    <div
      className="grid__col-6 margin-top--1 margin-bottom--1"
    >
      <form
        onSubmit={[Function]}
      >
        â—€ 
        <a
          onClick={[Function]}
          style={Object {}}
        >
          Back
        </a>
        <h2
          className="inline-block margin-left--1"
        >
          
        </h2>
        <button
          className="btn btn--primary btn--block margin-left--1"
          disabled={false}
        >
          Approve
        </button>
      </form>
    </div>
  </div>
  <iframe
    className="preview__iframe"
    id="iframe"
    src="///datasets/abc123/editions/2017/versions/2"
  />
</div>
`;
